### PR TITLE
Dynamic Model Provider

### PR DESCRIPTION
**Pull request by marcosmc**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
The dynamic-simulation command doesn't support mapping file.

**What is the new behavior (if this is a feature change)?**
The mapping file can be passed to the command, and loaded. Only groovy is supported yet.

### CODE REVIEW

File: dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/DynamicModelExtension.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+package com.powsybl.dynamicsimulation.groovy;
+
+import com.powsybl.dynamicsimulation.DynamicModel;
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public interface DynamicModelExtension extends GroovyExtension<DynamicModel> {
+}
```
Comment (agnesLeroy) [inline dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/DynamicModelExtension.java:16]: To be consistent with the curves this class should be named DynamicModelGroovyExtension

File: dynamic-simulation/dynamic-simulation-dsl/src/test/groovy/com/powsybl/dynamicsimulation/groovy/DummyDynamicModelExtension.groovy [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+                throw new DslException("'parameterSetId' field is not set")
+            }
+            if (!dynamicModelSpec.parameterSetFile) {
+                throw new DslException("'parameterSetFile' field is not set")
+            }
+
+            consumer.accept(new DummyDynamicModel(dynamicModelSpec.id, dynamicModelSpec.parameterSetId, dynamicModelSpec.parameterSetFile))
+        }
+    }
+}
```
Comment (agnesLeroy) [inline dynamic-simulation/dynamic-simulation-dsl/src/test/groovy/com/powsybl/dynamicsimulation/groovy/DummyDynamicModelExtension.groovy:60]: Same remark as above, this class should be named DummyDynamicModelGroovyExtension.

File: dynamic-simulation/dynamic-simulation-dsl/src/test/java/com/powsybl/dynamicsimulation/groovy/DummyDynamicModel.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public class DummyDynamicModel implements DynamicModel {
+
+    private final String id;
+
+    private final String parameterSetId;
+
+    private final String parameterSetFile;
```
Comment (agnesLeroy) [inline dynamic-simulation/dynamic-simulation-dsl/src/test/java/com/powsybl/dynamicsimulation/groovy/DummyDynamicModel.java:23]: Why do we have the parameterSetFile here? I was expecting to have only an Id (that will be either the static Id o the dynamic model Id) and a parameter Id. Am I wrong?

File: dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/DynamicModelExtension.java (hunk 1 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/DynamicModelExtension.java]: Done

File: dynamic-simulation/dynamic-simulation-dsl/src/test/java/com/powsybl/dynamicsimulation/groovy/DummyDynamicModel.java (hunk 3 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-dsl/src/test/java/com/powsybl/dynamicsimulation/groovy/DummyDynamicModel.java]: Done

File: dynamic-simulation/dynamic-simulation-dsl/src/test/groovy/com/powsybl/dynamicsimulation/groovy/DummyDynamicModelExtension.groovy (hunk 2 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-dsl/src/test/groovy/com/powsybl/dynamicsimulation/groovy/DummyDynamicModelExtension.groovy]: Done

File: dynamic-simulation/dynamic-simulation-dsl/src/test/resources/mapping.groovy [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+import com.powsybl.iidm.network.Generator
+import com.powsybl.iidm.network.Load
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+Logger logger = LoggerFactory.getLogger("com.powsybl.dynamicsimulation.groovy.GroovyCurvesProvider")
```
Comment (mathbagu) [inline dynamic-simulation/dynamic-simulation-dsl/src/test/resources/mapping.groovy:13]: You should adapt the logger name.

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java [hunk 5]
```diff
@@ -45,6 +48,7 @@
 public class DynamicSimulationTool implements Tool {
 
     private static final String CASE_FILE = "case-file";
+    private static final String MAPPING_FILE = "mapping-file";
```
Comment (mathbagu) [inline dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java:51]: we should also change this option's name

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java [hunk 6]
```diff
@@ -78,6 +82,12 @@ public Options getOptions() {
                     .argName("FILE")
                     .required()
                     .build());
+                options.addOption(Option.builder().longOpt(MAPPING_FILE)
+                    .desc("mapping description as Groovy file")
```
Comment (mathbagu) [inline dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java:86]: And give more detail in the description

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java [hunk 7]
```diff
@@ -131,6 +141,12 @@ public void run(CommandLine line, ToolRunningContext context) throws Exception {
 
         DynamicSimulation.Runner runner = DynamicSimulation.find();
 
+        MappingSupplier mappingSupplier = MappingSupplier.empty();
+        if (line.hasOption(MAPPING_FILE)) {
+            Path mappingFile = context.getFileSystem().getPath(line.getOptionValue(MAPPING_FILE));
+            mappingSupplier = createMappingSupplier(mappingFile, runner.getName());
+        }
```
Comment (mathbagu) [inline dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java:148]: This option is required so this test is useless.

File: dynamic-simulation/dynamic-simulation-api/src/main/java/com/powsybl/dynamicsimulation/MappingSupplier.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public interface MappingSupplier extends SimulatorInputSupplier<DynamicModel> {
+
+    static MappingSupplier empty() {
+        return network -> Collections.emptyList();
+    }
+}
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply dynamic-simulation/dynamic-simulation-api/src/main/java/com/powsybl/dynamicsimulation/MappingSupplier.java:20]: Yes I agree it should be moved to the test sources

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java (hunk 6 above)
Comment (agnesLeroy) [reply dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java:86]: Would "Dynamic models description as a Groovy file: defines the dynamic models to be associated to chosen equipments of the network" be enough?

File: dynamic-simulation/dynamic-simulation-api/src/main/java/com/powsybl/dynamicsimulation/MappingSupplier.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (zamarrenolm) [reply dynamic-simulation/dynamic-simulation-api/src/main/java/com/powsybl/dynamicsimulation/MappingSupplier.java:20]: Agree on the comments about the name of the class. @mathbagu suggestion of renaming from `MappingSupplier` to `DynamicModelsSupplier` is ok for me

[the remark this answers is not in the document]
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-api/src/main/java/com/powsybl/dynamicsimulation/MappingSupplier.java]: Done

File: dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/GroovyMappingSupplier.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+import com.powsybl.iidm.network.Network;
+
+import groovy.lang.Binding;
+import groovy.lang.GroovyCodeSource;
+import groovy.lang.GroovyShell;
+
+/**
+ * @author Marcos de Miguel <demiguelm at aia.es>
+ */
+public class GroovyMappingSupplier implements MappingSupplier {
```
[the remark this answers is not in the document]
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-dsl/src/main/java/com/powsybl/dynamicsimulation/groovy/GroovyMappingSupplier.java]: Done

File: dynamic-simulation/dynamic-simulation-dsl/src/test/resources/mapping.groovy (hunk 4 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-dsl/src/test/resources/mapping.groovy]: Done.
Should we change "GroovyCurvesProvider" to "GroovyCurvesSupplier" in curves.groovy?

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java (hunk 5 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java]: Done

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java (hunk 6 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java]: Description changed by Agnes proposal

File: dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java (hunk 7 above)
Comment (marcosmc) [reply dynamic-simulation/dynamic-simulation-tool/src/main/java/com/powsybl/dynamicsimulation/tool/DynamicSimulationTool.java]: Done

**Pull request merged**